zaza/openstack/charm_tests/ovn/setup.py

```diff
@@ -0,0 +1,139 @@
+        # endpoints
+        GENEVE_ENCAP_OVERHEAD = 38
+        IP4_HEADER_SIZE = 20
+        return int(n_api_config['global-physnet-mtu']['value']) - (
+            GENEVE_ENCAP_OVERHEAD + IP4_HEADER_SIZE)
+
+    def configure_ngw_novs(self):
+        """Configure n-ovs and n-gw units."""
+        cfg = {
+            # To be able to successfully clean up after the Neutron agents we
+            # need to use the 'openvswitch' `firewall-driver`.
+            'firewall-driver': 'openvswitch',
+            # To be able to have instances successfully survive the migration
+            # without communication issues we need to lower the MTU announced
+            # to instances prior to migration.
+            #
+            # NOTE: In a real world scenario the end user would configure the
+            # MTU at least 24 hrs prior to doing the migration to allow
+            # instances to reconfigure as they renew the DHCP lease.
+            #
+            # NOTE: For classic n-gw topologies the `instance-mtu` config
+            # is a NOOP on neutron-openvswitch units, but that is ok.
+            'instance-mtu': self._get_instance_mtu_from_global_physnet_mtu()
+        }
+        apps = ('neutron-gateway', 'neutron-openvswitch')
+        for app in apps:
+            try:
+                zaza.model.get_application(app)
+                for k, v in cfg.items():
+                    logging.info('Setting `{}` to "{}" on "{}"...'
+                                 .format(k, v, app))
+                with self.config_change(cfg, cfg, app):
+                    # The intent here is to change the config and not restore
+                    # it. We accomplish that by passing in the same value for
+                    # default and alternate.
+                    #
+                    # The reason for using the `config_change` helper for this
+                    # is that it already deals with all the permutations of
+                    # config already being set etc and does not get into
+                    # trouble if the test bundle already have the values we try
+                    # to set.
+                    continue
+            except KeyError:
+                pass
+
+    def configure_ovn_mappings(self):
+        """Copy mappings from n-gw or n-ovs application."""
+        dst_apps = ('ovn-dedicated-chassis', 'ovn-chassis')
+        src_apps = ('neutron-gateway', 'neutron-openvswitch')
+        ovn_cfg = {}
+        for app in src_apps:
+            try:
+                app_cfg = zaza.model.get_application_config(app)
+                ovn_cfg['bridge-interface-mappings'] = app_cfg[
+                    'data-port']['value']
+                ovn_cfg['ovn-bridge-mappings'] = app_cfg[
+                    'bridge-mappings']['value']
+                # Use values from neutron-gateway when present, otherwise use
+                # values from neutron-openvswitch
+                break
+            except KeyError:
```

Comment:
You have used this pattern multiple times now. It might make sense to create a method that takes in a list of applications and a list of config parameters and returns a dictionary of config data.

Reply:
Encapsulating patterns in helper functions is a good thing, but while they look similar all three uses here are different.

This section blindly sets configuration on all listed apps whether they are present or not:
https://github.com/openstack-charmers/zaza-openstack-tests/blob/5a10779e4533cb38d42a15f01923789769aeb683/zaza/openstack/charm_tests/ovn/setup.py#L61-L80

This section consciously sets configuration on _one_ specific app from a set of alternatives and fails if none of them are there:
https://github.com/openstack-charmers/zaza-openstack-tests/blob/5a10779e4533cb38d42a15f01923789769aeb683/zaza/openstack/charm_tests/ovn/setup.py#L105-L121

This section _retrieves_ and stores configuration from the listed apps:
https://github.com/openstack-charmers/zaza-openstack-tests/blob/5a10779e4533cb38d42a15f01923789769aeb683/zaza/openstack/charm_tests/ovn/setup.py#L87-L103

Reply:
I collapsed the two first use cases into a helper method that takes an argument to choose operation mode.

Reply:
Got it. I had a superficial view. No need to collapse them if that is not beneficial.